src/test/scala/com/dwolla/mysql/init/RoundTripIntegrationTest.scala

```diff
@@ -0,0 +1,212 @@
+import munit._
+import natchez.Span
+import org.http4s.client.Client
+import org.http4s.dsl.Http4sDsl
+import org.http4s.{HttpApp, Request}
+import org.scalacheck.Arbitrary.arbitrary
+import org.scalacheck.effect.PropF
+import org.scalacheck.util.Pretty
+import org.scalacheck.{Arbitrary, Gen}
+import org.typelevel.jawn.Parser
+import org.typelevel.log4cats.Logger
+import _root_.io.circe.optics.JsonPath._
+
+import scala.concurrent.duration._
+
+class RoundTripIntegrationTest
+  extends CatsEffectSuite
+    with ScalaCheckEffectSuite
+    with CatsEffectFixtures
+    with CloudFormationCustomResourceArbitraries
+    with ArbitraryRefinedTypes {
+
+  override def munitTimeout: Duration = 10.minutes
+
+  private def genResources[F[_] : Monad, A](max: Int = 10)
+                                           (implicit R: Arbitrary[Resource[F, A]]): Gen[Resource[F, List[A]]] =
+    for {
+      size <- Gen.chooseNum(0, max)
+      resources <- Gen.listOfN(size, R.arbitrary)
+    } yield resources.sequence
+
+  private def genSecret[F[_] : Functor, A : Encoder : Arbitrary](secretsAlg: SecretsManagerAlg[F]): Gen[Resource[F, SecretId]] =
+    for {
+      nameSuffix <- Gen.stringOfN(20, Gen.oneOf(Gen.alphaNumChar, Gen.oneOf('-', '/', '_', '+', '=', '.', '@', '!')))
+      a <- arbitrary[A]
+      name <- refineV[SecretNamePredicate](s"integrationTestSecret-$nameSuffix").fold(_ => Gen.fail, Gen.const)
+    } yield Resource.make(secretsAlg.createJsonSecret(name, a))(secretsAlg.deleteSecret(_, Immediate))
+
+  private val genUserConnectionInfo: Option[Database] => Gen[UserConnectionInfo] = maybeDatabase =>
+    for {
+      database <- maybeDatabase.fold(arbitrary[SqlIdentifier].map(Database(_)))(Gen.const)
+      host <- Gen.const(Host("localhost"))
+      port <- Gen.const(Port(3306))
+      user <- arbitrary[MySqlUser].map(Username(_))
+      password <- arbitrary[GeneratedPassword].map(Password(_))
+    } yield UserConnectionInfo(database, host, port, user, password)
+
+  private def genDatabaseMetadata[F[_] : Monad, A : Encoder](secretsAlg: SecretsManagerAlg[F],
+                                                             genA: Option[Database] => Gen[A]): Gen[Resource[F, DatabaseMetadata]] = {
+    for {
+      host <- Gen.const(Host("localhost"))
+      port <- Gen.const(Port(3306))
+      database <- arbitrary[SqlIdentifier].map(Database(_))
+      username <- Gen.const[MySqlUser]("root").map(MasterDatabaseUsername(_))
+      password <- Gen.const("password").map(MasterDatabasePassword(_))
+      secrets <- {
+        implicit val arbA: Arbitrary[A] = Arbitrary(genA(database.some))
+        implicit val arbSecret: Arbitrary[Resource[F, SecretId]] = Arbitrary(genSecret[F, A](secretsAlg))
+
+        genResources[F, SecretId]()
+      }
+    } yield secrets.map(DatabaseMetadata(host, port, database, username, password, _))
+  }
+
+  /**
+   * Secrets Manager secrets cost $0.40 / month, so it could be expensive to run this using the real
```

Comment:
This isn't necessary I don't think, but one thing I wondered about: would it be possible to use something like [localstack](https://localstack.cloud/) to run this round-trip property test against a somewhat more "real" instance of SecretsManager?

Reply:
It might be, but unless we start having issues with the implementation of `SecretsManagerAlg` I'd prefer to hold off on that for now.